File: mcp-list/packages/EventAnalyzer/src/api_client.py
"""API Client
调用埋点接口获取事件字段定义
"""

import json
import requests
from typing import Dict, Any, Optional
from functools import lru_cache


class EventAPIClient:
    """埋点事件 API 客户端"""

    BASE_URL = "https://tptest-3d66.top/trans/api/event"

    def __init__(self, timeout: int = 10):
        """
        初始化 API 客户端

        Args:
            timeout: 请求超时时间（秒）
        """
        self.timeout = timeout
        self.session = requests.Session()

    @lru_cache(maxsize=128)
    def get_event_fields(self, event_name: str) -> Dict[str, Any]:
        """
        获取事件的所有字段定义（带缓存）

        Args:
            event_name: 事件名称（如 LlwResExposure）

        Returns:
            字段定义字典
            格式: {
                "field_name": {
                    "type": "NUMBER/STRING/BOOL/LIST",
                    "tips": "字段说明",
                    "desc": "详细描述",
                    "trans": "枚举值映射 JSON 字符串"
                }
            }

        Raises:
            Exception: 请求失败时抛出异常
        """
        try:
            response = self.session.get(
                self.BASE_URL,
                params={"event": event_name},
                timeout=self.timeout
            )
            response.raise_for_status()
            return response.json()

        except requests.exceptions.RequestException as e:
            raise Exception(f"获取事件字段定义失败: {str(e)}")

    def parse_field_trans(self, trans_str: str) -> Dict[str, str]:
        """
        解析 trans 字段的枚举值映射

        Args:
            trans_str: trans 字段的 JSON 字符串

        Returns:
            枚举值映射字典
        """
        if not trans_str:
            return {}

        try:
            return json.loads(trans_str)
        except json.JSONDecodeError:
            return {}

    def get_field_info(self, event_name: str, field_name: str) -> Optional[Dict[str, Any]]:
        """
        获取单个字段的信息

        Args:
            event_name: 事件名称
            field_name: 字段名称

        Returns:
            字段信息，如果字段不存在返回 None
        """
        fields = self.get_event_fields(event_name)
        field_info = fields.get(field_name)

        if field_info and field_info.get("trans"):
            # 解析枚举值
            field_info["enum_values"] = self.parse_field_trans(field_info["trans"])

        return field_info

    def get_all_field_names(self, event_name: str) -> list[str]:
        """
        获取事件的所有字段名称列表

        Args:
            event_name: 事件名称

        Returns:
            字段名称列表
        """
        fields = self.get_event_fields(event_name)
        return list(fields.keys())

    def clear_cache(self):
        """清空缓存"""
        self.get_event_fields.cache_clear()

File: mcp-list/packages/EventAnalyzer-node/src/services/apiClient.ts
/**
 * 埋点事件 API 客户端
 * 负责调用 API 获取事件字段定义
 */

import axios, { AxiosInstance } from 'axios';
import NodeCache from 'node-cache';

export interface FieldDefinition {
  type: string;
  desc: string;
  trans?: string;
  required?: boolean;
  enum_values?: Record<string, string>;
}

export interface FieldDefinitions {
  [fieldName: string]: FieldDefinition;
}

export class EventAPIClient {
  private baseUrl: string;
  private cache: NodeCache;
  private axiosInstance: AxiosInstance;

  constructor() {
    // 从环境变量获取 API 地址
    this.baseUrl = process.env.EVENT_API_BASE_URL || 'https://tptest-3d66.top/trans/api/event';

    // 初始化缓存（TTL 1 小时）
    this.cache = new NodeCache({ stdTTL: 3600 });

    // 初始化 axios 实例
    this.axiosInstance = axios.create({
      timeout: 10000,
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }

  /**
   * 获取事件字段定义
   * @param eventName 事件名称
   */
  async getEventFields(eventName: string): Promise<FieldDefinitions> {
    // 检查缓存
    const cacheKey = `event:${eventName}`;
    const cached = this.cache.get<FieldDefinitions>(cacheKey);
    if (cached) {
      return cached;
    }

    try {
      // 调用 API
      const response = await this.axiosInstance.get(this.baseUrl, {
        params: { event: eventName },
      });

      const fields: FieldDefinitions = response.data;

      // 缓存结果
      this.cache.set(cacheKey, fields);

      return fields;
    } catch (error: any) {
      throw new Error(`获取事件字段定义失败: ${error.message}`);
    }
  }

  /**
   * 解析字段的枚举值字符串
   * @param transStr 枚举值字符串，格式：0:未知,1:成功,2:失败
   * @returns 枚举值映射对象 { "0": "未知", "1": "成功", "2": "失败" }
   */
  parseFieldTrans(transStr: string): Record<string, string> {
    if (!transStr) {
      return {};
    }

    const result: Record<string, string> = {};

    try {
      // 分割枚举项
      const items = transStr.split(',');
      for (const item of items) {
        const [key, value] = item.split(':');
        if (key && value) {
          result[key.trim()] = value.trim();
        }
      }
    } catch {
      // 解析失败，返回空对象
    }

    return result;
  }

  /**
   * 获取单个字段的信息
   * @param eventName 事件名称
   * @param fieldName 字段名称
   */
  async getFieldInfo(eventName: string, fieldName: string): Promise<FieldDefinition | null> {
    const fields = await this.getEventFields(eventName);
    return fields[fieldName] || null;
  }

  /**
   * 清除缓存
   */
  clearCache(): void {
    this.cache.flushAll();
  }

  /**
   * 清除特定事件的缓存
   */
  clearEventCache(eventName: string): void {
    this.cache.del(`event:${eventName}`);
  }
}
